# Bioinformatics-Workshop

## [Progect 1. What causes antibiotic resistance?”. Alignment to reference, variant calling.](https://github.com/LiliiaBgdnv/Bioinformatics-Workshop/tree/Project_1)
The sequencing data were analyzed and the mutations responsible for giving E. coli antibiotic resistance properties were found. 

**Skills:**
> * FastQC
> * Trimmimatic
> * Aligning sequences to reference
> * Variant calling
> * SNP annotation

## [Progect 2. Why did I get the flu?](https://github.com/LiliiaBgdnv/Bioinformatics-Workshop/tree/Progect-2)
 In this work we analyze sequencing data from a person who was infected with the human influenza virus H3N2.
Afterwards, we identified the mutation that resulted in an amino acid substitution in a protein known as the epitope, which is recognized by
antibodies. Therefore, SNP was found at position 307, epitope D. Proline was switched out for serine as a result of
the mutation (P103S), which could have led to breakthrough infection.

**Skills:**
> * Deep sequence data
> * awk one-liners

## [Project 3. "E. coli outbreak investigation". De novo assembly and annotation of bacterial genomes.](https://github.com/LiliiaBgdnv/Bioinformatics-Workshop/tree/Progect_3)
In this paper, we studied how E.coli could lead to an epidemic in Germany in 2011 that caused dozens
of deaths. For this purpose we have taken the steps of assembling the genome of the E.coli X strain,
comparing it to another E.coli strain, determining which genes our strain has acquired and how it became
pathogenic. During the work we found out that strain E.coli X acquired the genes stxB and stxA from the
phage, which led to the ability to secrete shiga toxin. We also found the bla_1 and bla_2 genes encoding
beta-lactamase in the CTX-M family and TEM family, respectively.

**Skills:**
> * K-mer profile
> * Assembling genome
> * Genome Annotation
